cli/flags.rs

```diff
@@ -1314,41 +1316,96 @@ pub fn resolve_urls(urls: Vec<String>) -> Vec<String> {
+  }
+
   let mut out: Vec<String> = vec![];
   for host_and_port in paths.iter() {
+    // pure port:                 :1234
+    // ipv4 with port:            1.2.3.4:1234
+    // ipv6 with port:            [1:2:3::4]:1234
+    // pure ipv4:                 1.2.3.4
+    // pure ipv6 with bracket:    [1:2:3::4]
+    // pure ipv6 without bracket: 1:2:3::4
+    // pure domain name:          example.com
+    // domain name with port:     example.com:1234
+
+    if host_and_port.parse::<SocketAddr>().is_ok() {
+      // ipv4 with port / ipv6 with port
+      out.push(host_and_port.to_owned());
+      continue;
+    }
+    if host_and_port.parse::<IpAddr>().is_ok() {
+      // pure ipv4 / pure ipv6 without bracket
+      out.push(host_and_port.to_owned());
+      continue;
+    }
+    if host_and_port.starts_with('[')
+      && host_and_port.ends_with(']')
+      && host_and_port[1..host_and_port.len() - 1]
+        .parse::<Ipv6Addr>()
+        .is_ok()
+    {
+      // pure ipv6 with bracket
+      out.push(host_and_port.to_owned());
+      continue;
+    }
+    if host_and_port.starts_with(':') {
+      let port = host_and_port[1..].parse::<i32>().unwrap_or(0);
+      if let Ok(hosts) = expand_port_to_localhost(port) {
+        out.extend(hosts);
+        continue;
+      }
+    }
     let parts = host_and_port.split(':').collect::<Vec<&str>>();
-
-    match parts.len() {
-      // host only
-      1 => {
+    if !parts.is_empty()
+      && parts.len() < 3
+      && DOMAIN_ALLOWED_CHARACTERS.is_match(parts[0])
+    {
+      if parts.len() == 1 {
+        // pure domain name
         out.push(host_and_port.to_owned());
-      }
-      // host and port (NOTE: host might be empty string)
-      2 => {
-        let host = parts[0];
-        let port = parts[1];
-
-        if !host.is_empty() {
+        continue;
+      } else if parts.len() == 2 {
+        let port = parts[1].parse::<i32>().unwrap_or(0);
+        if port > 0 && port < 65536 {
+          // domain name with port
           out.push(host_and_port.to_owned());
           continue;
         }
-
-        // we got bare port, let's add default hosts
-        for host in ["0.0.0.0", "127.0.0.1", "localhost"].iter() {
-          out.push(format!("{}:{}", host, port));
-        }
       }
-      _ => panic!("Bad host:port pair: {}", host_and_port),
     }
+    eprintln!(
+      "{}: Bad host or host:port pair: {}",
+      crate::colors::yellow("Warning".to_string()),
+      host_and_port
+    );
   }
 
   out
 }
 
+/// Expands port into full localhost paths.
+/// It expands port into 5 paths with following hosts: `0.0.0.0:port`,
+/// `127.0.0.1:port`, `[::]:port`, `[::1]:port` and `localhost:port`.
+fn expand_port_to_localhost(port: i32) -> Result<Vec<String>, &'static str> {
+  if port <= 0 || port >= 65536 {
+    return Err("invalid port");
+  }
+
+  let mut out: Vec<String> = vec![];
+
+  for host in ["0.0.0.0", "127.0.0.1", "[::]", "[::1]", "localhost"].iter() {
```

Comment:
I don't think "localhost" should be included in this list. It is not necessarily equivalent to "127.0.0.1" or "::1".

Also I think the error returned from this is unnecessary. I suggest returning`Option<Vec<String>>` or using something from `cli/op_error.rs`. I suggest the former, since the error here will not be exposed to the JS runtime.

Reply:
Actually I'm confusing about this code, is it necessary to expand the port to IP addresses?
Think about this code:
```typescript
import { serve } from "https://deno.land/std@0.50.0/http/server.ts";
const server = serve(":8080");
for await (const req of server) {
  req.respond({ body: "Hello World\n" });
}
```
It required permission `:8080`, not any of `127.0.0.1:8080`, `0.0.0.0:8080`, `[::1]:8080` or `[::]:8080`.

Any idea, should we just keep origin inputs `:port`?